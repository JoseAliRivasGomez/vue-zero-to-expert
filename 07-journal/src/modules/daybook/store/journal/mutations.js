

export const setEntries = (state, entries) => {
    state.entries = [...state.entries, ...entries]
    state.isLoading = false
}

export const updateEntry = (state, entry) => {
    const idx = state.entries.map(e => e.id).indexOf(entry.id)
    state.entries[idx] = entry
    //console.log({idx})
}

export const createEntry = (state, entry) => {
    state.entries = [entry, ...state.entries]
}

export const deleteEntry = (state, id) => {
    state.entries = state.entries.filter(e => e.id != id)
}