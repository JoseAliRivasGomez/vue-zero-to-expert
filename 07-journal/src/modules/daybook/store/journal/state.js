

export default () => ({
    isLoading: true,
    entries: [
        {
            id: '1',
            date: new Date().toDateString(),
            text: 'Veniam enim consectetur voluptate ad occaecat occaecat proident veniam esse adipisicing. Mollit fugiat amet quis ad elit enim nulla exercitation eu exercitation. Mollit sit consectetur deserunt ullamco enim cupidatat est. Irure et consectetur elit occaecat ullamco nulla quis.',
            picture: null,
        },
        {
            id: '2',
            date: new Date().toDateString(),
            text: 'Adipisicing irure exercitation ipsum id fugiat sit occaecat culpa cupidatat. Elit labore exercitation duis ea dolor laboris duis ea Lorem minim. Cillum magna mollit eiusmod nulla consectetur nostrud est.',
            picture: null,
        },
        {
            id: '3',
            date: new Date().toDateString(),
            text: 'Ut Lorem consequat pariatur sit cillum deserunt. Dolore est cupidatat officia labore aliquip veniam tempor nostrud sunt velit enim cupidatat. Dolore in consequat eiusmod ea elit mollit ex sunt voluptate nulla consectetur cillum. Officia sunt proident sint aliquip occaecat nulla elit esse anim adipisicing cupidatat dolor. Ullamco aliqua labore dolor fugiat cupidatat officia officia culpa. Ea adipisicing excepteur minim consectetur ad minim.',
            picture: null,
        },
    ]
})