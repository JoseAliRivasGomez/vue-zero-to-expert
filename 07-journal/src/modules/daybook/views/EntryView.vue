<template>
    <template v-if="entry">
    <div  class="entry-title d-flex justify-content-between p-2">
        <div>
            <span class="text-success fs-3 fw-bold">{{day}}</span>
            <span class="mx-1 fs-3">{{month}}</span>
            <span class="mx-2 fs-4 fw-light">{{yearDay}}</span>
        </div>
        <div>
            <input type="file" @change="onSelectedImage" ref="imageSelector" v-show="false" accept="image/png, image/jpeg">
            <button v-if="entry.id" class="btn btn-danger mx-2" @click="onDeleteEntry">
                Borrar
                <i class="fa fa-trash-alt"></i>
            </button>
            <button class="btn btn-primary" @click="onSelectImage">
                Subir foto
                <i class="fa fa-upload"></i>
            </button>
        </div>
    </div>
    <hr>
    <div class="d-flex flex-column px-3 h-75">
        <textarea v-model="entry.text" placeholder="Que paso hoy?"></textarea>
    </div>
        <img v-if="entry.picture && !localImage" :src="entry.picture" alt="Entry Picture" class="img-thumbnail">
        <img v-if="localImage" :src="localImage" alt="Entry Picture" class="img-thumbnail">
    </template>
    <Fab icon="fa-save" @on:click="saveEntry" />
</template>

<script>
import { defineAsyncComponent } from 'vue'
import { mapGetters, mapActions } from 'vuex'
import getDayMonthYear from "../helpers/getDayMonthYear";
import uploadImage from "../helpers/uploadImage";
import Swal from 'sweetalert2'


export default {
    props: {
        id: {
            type: String,
            required: true
        }
    },
    components: {
        Fab: defineAsyncComponent(() => import('@/modules/daybook/components/Fab.vue'))
    },
    data(){
        return{
            entry: null,
            localImage: null,
            file: null
        }
    },
    computed: {
        ...mapGetters('journal', ['getEntryById']),
        day(){
            const {day} = getDayMonthYear(this.entry.date)
            return day
        },
        month(){
            const {month} = getDayMonthYear(this.entry.date)
            return month
        },
        yearDay(){
            const {yearDay} = getDayMonthYear(this.entry.date)
            return yearDay
        },
    },
    methods: {
        ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry']),
        loadEntry() {

            this.localImage = null
            this.file = null

            let entry
            if(this.id === 'new'){
                entry = {
                    text: '',
                    date: new Date().getTime()
                }
            }else{
                entry = this.getEntryById(this.id)
                if(!entry) return this.$router.push({name: 'no-entry'})
            }

            this.entry = entry
        },
        async saveEntry(){

            new Swal({
                title: 'Espere por favor',
                allowOutsideClick: false
            })
            Swal.showLoading()

            if(this.file){
                const imageURL = await uploadImage(this.file)
                this.entry.picture = imageURL
            }
            
            if(this.entry.id){
                //console.log('update');
                await this.updateEntry(this.entry)
            }else{
                //console.log('create');
                const id = await this.createEntry(this.entry)

                this.$router.push({name: 'entry', params: {id}})
            }

            Swal.fire('Guardado', 'Entrada guardada con exito', 'success')
            this.file = null
        },
        async onDeleteEntry(){
            const {isConfirmed} = await Swal.fire({
                title: 'Esta seguro?',
                text: 'Una vez eliminado, no se puede recuperar',
                showDenyButton: true,
                confirmButtonText: 'Si, estoy seguro'
            })

            if(isConfirmed){
                new Swal({
                    title: 'Espere por favor',
                    allowOutsideClick: false
                })
                Swal.showLoading()

                await this.deleteEntry(this.entry.id)
                this.$router.push({name: 'no-entry'})

                Swal.fire('Eliminado', 'Entrada eliminada con exito', 'success')
            }
            
        },
        onSelectedImage(event){
            const file = event.target.files[0]
            if(!file){
                this.localImage = null
                this.file = null
                return
            }

            this.file = file

            const fr = new FileReader()
            fr.onload = () => this.localImage = fr.result
            fr.readAsDataURL(file)

            
        },
        onSelectImage(){
            this.$refs.imageSelector.click()
        }
    },
    created(){
        this.loadEntry()
    },
    watch: {
        id(value, oldValue){
            this.loadEntry()
        }
    }

}
</script>

<style lang="scss" scoped>

textarea{
    font-size: 20px;
    border: none;
    height: 100%;

    &:focus{
        outline: none;
    }
}

img{
    width: 200px;
    position: fixed;
    bottom: 150px;
    right: 20px;
    box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2)
}

</style>