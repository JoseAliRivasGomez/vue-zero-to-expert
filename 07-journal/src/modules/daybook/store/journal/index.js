

import state from './state'
import * as actions from './actions'
import * as mutations from './mutations'
import * as getters from './getters'

const journalModule = {
    namespaced: true,
    state,
    mutations,
    actions,
    getters
}

export default journalModule