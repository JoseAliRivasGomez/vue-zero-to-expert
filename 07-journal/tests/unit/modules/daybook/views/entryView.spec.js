
import { journalState } from "../../../mock-data/test-journal-state";
import { shallowMount } from "@vue/test-utils";
import EntryView from '@/modules/daybook/views/EntryView.vue'
import { createStore } from "vuex";
import journal from "@/modules/daybook/store/journal";
import Swal from "sweetalert2";

const createVuexStore = (initialState) => createStore({
    modules: {
        journal: {
            ...journal,
            state: {...initialState}
        }
    }
})

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('EntryView', () => { 

    const store = createVuexStore(journalState)
    store.dispatch = jest.fn() //Para no borrar datos en la DB

    const mockRouter = {
        push: jest.fn()
    }

    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-NSEsuO2kOUZd-Oy0Zbc'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })

    test('Debe de sacar al usuario porque el id no existe', () => { 

        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'No existe'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store],
            }
        })

        expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'})

    })

    test('Debe de mostrar la entrada correctamente', () => { 

        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled()

    })

    test('Debe de borrar la entrada y salir', (done) => { 

        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true}))

        wrapper.find('.btn-danger').trigger('click')

        expect(Swal.fire).toHaveBeenCalledWith({
            title: 'Esta seguro?',
            text: 'Una vez eliminado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Si, estoy seguro'
        })

        setTimeout(() => {
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-NSEsuO2kOUZd-Oy0Zbc')
            expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'})
            done()
        }, 1000);
    

    })

})