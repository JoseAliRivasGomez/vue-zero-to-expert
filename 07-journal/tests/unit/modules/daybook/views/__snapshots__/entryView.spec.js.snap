// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryView Debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">4</span><span class="mx-1 fs-3">Abril</span><span class="mx-2 fs-4 fw-light">2023, Martes</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Que paso hoy?"></textarea></div>
<img src="https://res.cloudinary.com/dqf5buxhi/image/upload/v1680672780/journal-vue/cxa0fo0zdzizc10dgq7e.jpg" alt="Entry Picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
